ELF: Add branch-to-branch optimization.

pcc opened the pull request:
When code calls a function which then immediately tail calls another
function there is no need to go via the intermediate function. By
branching directly to the target function we reduce the program's working
set for a slight increase in runtime performance.

Normally it is relatively uncommon to have functions that just tail call
another function, but with LLVM control flow integrity we have jump tables
that replace the function itself as the canonical address. As a result,
when a function address is taken and called directly, for example after
a compiler optimization resolves the indirect call, or if code built
without control flow integrity calls the function, the call will go via
the jump table.

The impact of this optimization was measured using a large internal
Google benchmark. The results were as follows:

CFI enabled:  +0.1% ± 0.05% queries per second
CFI disabled: +0.01% queries per second [not statistically significant]

The optimization is enabled by default at -O2 but may also be enabled
or disabled individually with --{,no-}branch-to-branch.

This optimization is implemented for AArch64 and X86_64 only.

lld's runtime performance (real execution time) after adding this
optimization was measured using firefox-x64 from lld-speed-test [1]
with ldflags "-O2 -S" on an Apple M2 Ultra. The results are as follows:

```
    N           Min           Max        Median           Avg        Stddev
x 512     1.2264546     1.3481076     1.2970261     1.2965788   0.018620888
+ 512     1.2561196     1.3839965     1.3214632     1.3209327   0.019443971
Difference at 95.0% confidence
	0.0243538 +/- 0.00233202
	1.87831% +/- 0.179859%
	(Student's t, pooled s = 0.0190369)
```

[1] https://discourse.llvm.org/t/improving-the-reproducibility-of-linker-benchmarking/86057

pcc committed:
ELF: Add branch-to-branch optimization.

When code calls a function which then immediately tail calls another
function there is no need to go via the intermediate function. By
branching directly to the target function we reduce the program's working
set for a slight increase in runtime performance.

Normally it is relatively uncommon to have functions that just tail call
another function, but with LLVM control flow integrity we have jump tables
that replace the function itself as the canonical address. As a result,
when a function address is taken and called directly, for example after
a compiler optimization resolves the indirect call, or if code built
without control flow integrity calls the function, the call will go via
the jump table.

The impact of this optimization was measured using a large internal
Google benchmark. The results were as follows:

CFI enabled:  +0.1% ± 0.05% queries per second
CFI disabled: +0.01% queries per second [not statistically significant]

The optimization is enabled by default at -O2 but may also be enabled
or disabled individually with --{,no-}branch-to-branch.

This optimization is implemented for AArch64 and X86_64 only.

lld's runtime performance (real execution time) after adding this
optimization was measured using firefox-x64 from lld-speed-test [1]
with ldflags "-O2 -S" on an Apple M2 Ultra. The results are as follows:

```
    N           Min           Max        Median           Avg        Stddev
x 512     1.2264546     1.3481076     1.2970261     1.2965788   0.018620888
+ 512     1.2561196     1.3839965     1.3214632     1.3209327   0.019443971
Difference at 95.0% confidence
	0.0243538 +/- 0.00233202
	1.87831% +/- 0.179859%
	(Student's t, pooled s = 0.0190369)
```

[1] https://discourse.llvm.org/t/improving-the-reproducibility-of-linker-benchmarking/86057

Pull Request: https://github.com/llvm/llvm-project/pull/138366

MaskRay commented on lld/ELF/Options.td:
```diff
@@ -59,6 +59,10 @@ def build_id: J<"build-id=">, HelpText<"Generate build ID note">,
   MetaVarName<"[fast,md5,sha1,uuid,0x<hexstring>]">;
 def : F<"build-id">, Alias<build_id>, AliasArgs<["sha1"]>, HelpText<"Alias for --build-id=sha1">;
 
+defm branch_to_branch: B<"branch-to-branch",
```
BB (only allow `--` long options) for new options (no GNU ld compatibility concern)

MaskRay commented on lld/test/ELF/aarch64-branch-to-branch.s:
```diff
@@ -0,0 +1,58 @@
+vtable:
+# B2B: Contents of section .rodata:
+# B2B-NEXT: [[VF:[0-9a-f]{8}]]
+.4byte f1@PLT - vtable
```
Note: `f1@PLT - .` is unofficial syntax (in both AArch64 and X86) that will change. It should be straightforward to update the uses when the change happens.

(For RISC-V I've switched to `%pltpcrel(foo+offset)` (`#132569`)

MaskRay commented on lld/test/ELF/aarch64-branch-to-branch.s:
```diff
@@ -0,0 +1,58 @@
+# B2B: b {{.*}} <f3>
+# NOB2B: b {{.*}} <f2{{.*}}>
+b f2
+
```
There is no negative test (when the destination does not start with a branch instruction)

MaskRay reviewed (commented):
(Still on a trip with limited computer access)

smithp35 reviewed (commented):
As a general idea this looks good to me. Some thoughts on some possible additional tests and edge-cases.

The bolt folk have been asking if the range-extension thunks can output relocations so that they can recreate the control flow more easily. It looks like this will handle emit-relocs naturally, could be worth a test case to check.

This looks like it runs before range extension thunks, which likely simplifies the logic as we don't need to check for branch range. Due to the way thunks are currently placed it could be possible that we end up removing a branch to a branch, only for the thunk generation code placing a range extension thunk such that a long-branch is generated. For example:
```
fn:
b fn2
...
fn2: // in range of fn
b fn3
...
fn3: // in range of fn2 but not fn
``` 
If we change that so that fn branches directly to fn3 our range extension thunk may not be conveniently placed, so that an indirect branch is needed.

If this is a concern for large programs it may be better to do this after range extension thunks, although it would need a range check before applying it.

Alternately running it earlier than garbage collection would in theory permit some sections (branch in its own section) to be removed, but that's likely to be a neglible saving.

pcc commented:
> The bolt folk have been asking if the range-extension thunks can output relocations so that they can recreate the control flow more easily. It looks like this will handle emit-relocs naturally, could be worth a test case to check.

It doesn't look like that works because InputSection::copyRelocations reads the original relocation section and not the relocations array.

> This looks like it runs before range extension thunks, which likely simplifies the logic as we don't need to check for branch range. Due to the way thunks are currently placed it could be possible that we end up removing a branch to a branch, only for the thunk generation code placing a range extension thunk such that a long-branch is generated.

Correct. I placed it before range extension thunks deliberately, as it means that in some cases we would be able to avoid the range extension thunk entirely. I had the CFI jump table case in mind, where typically all the jump tables will appear near the end of the program (link position of the combined full LTO object file). For example:
```
f1:
b f2
f2.cfi:
ret
[....]
f2:
b f2.cfi
```
So it's neutral (or possibly positive) in terms of the number of branches executed, but possibly negative for code size. That being said, this optimization did not change the number of thunks created in lld-speed-test/firefox-arm64, but that doesn't use CFI. (The optimization didn't fire at all in lld-speed-test/chrome because of BTI instructions, but that should be fixable in a followup.)

> Alternately running it earlier than garbage collection would in theory permit some sections (branch in its own section) to be removed, but that's likely to be a neglible saving.

Correct. Also, if it ran before ICF that could expose more ICF opportunities but that's likely to be minor as well.

I considered placing it before ICF and GC, but we don't really have any infrastructure for rewriting relocation targets until reaching the relocation scan in the writer, and if we moved that earlier it would mean doing extra work (creating relocation lists for ICF'd and GC'd sections) for only a minor benefit.

pcc commented:
#138795 fixes the MC bug cited above. I'll implement the old object file workaround later.

pcc commented:
I changed my mind about https://github.com/llvm/llvm-project/pull/138366#issuecomment-2855889424 and decided that the LLVM assembler is correct and it's a linker bug that this case isn't handled.

I also updated this PR with two more bug fixes found during testing:
- Don't follow relocations to STT_GNU_IFUNC symbols
- Don't get stuck in an infinite loop when encountering a branch loop

pcc commented:
Ping.

smithp35 reviewed (commented):
I've made a few suggestions for some names that I found a bit confusing. The AArch64 side looks good to me. I've not checked the x86_64 side.

pcc commented on lld/ELF/Arch/TargetImpl.h:
```diff
@@ -0,0 +1,92 @@
+// std::nullopt. The effective addend for a relocation is the addend that is
+// used to determine its branch destination.
+//
+// getBranchInfo: If a control transfer relocation referring to is+offset
```
Done

pcc commented on lld/ELF/Arch/TargetImpl.h:
```diff
@@ -0,0 +1,92 @@
+// section, return the relocation for the branch target as well as its effective
+// addend (see above). Otherwise return {nullptr, 0}.
+//
+// mergeControlTransferRelocations: Given r1, a relocation for which
```
Done

rlavaee commented:
At my first glance, it seems this feature overwrites the target of the branch and doesn't move or relax the branches within the section, which would interfere with the SHT_LLVM_BB_ADDR_MAP. Could you please add a test just to be safe? You could use amend  a test with the SHT_LLVM_BB_ADDR_MAP section (https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/X86/basic-block-address-map.ll).

pcc commented:
Right, this feature doesn't change section sizes, so there shouldn't be an interaction with SHT_LLVM_BB_ADDR_MAP. AFAICT LLD doesn't contain code that parses SHT_LLVM_BB_ADDR_MAP so I don't see value in adding a test for it.

smithp35 commented on lld/ELF/Arch/AArch64.cpp:
```diff
@@ -975,6 +977,62 @@ void AArch64::relocateAlloc(InputSectionBase &sec, uint8_t *buf) const {
+  if (i != is.relocations.end() && i->offset == offset &&
+      i->type == R_AARCH64_JUMP26) {
+    return {i, i->addend};
+  }
```
Agree that BTI instructions should be in a separate patch. It would require disassembling to find one so may result in longer link times. Skipping over BTI with direct branches could apply even when the target wasn't another branch.

smithp35 commented:
Thanks for the updates. I don't have any more comments.

MaskRay commented:
> [pcc](https://github.com/pcc) wants to merge 1 commit into [users/pcc/spr/main.elf-add-branch-to-branch-optimization](https://github.com/llvm/llvm-project/tree/users/pcc/spr/main.elf-add-branch-to-branch-optimization) from [users/pcc/spr/elf-add-branch-to-branch-optimization](https://github.com/llvm/llvm-project/tree/users/pcc/spr/elf-add-branch-to-branch-optimization)

Is the base branch associated with an open PR? 

Neither
```
curl -L https://github.com/llvm/llvm-project/pull/138366.diff | patch -p1

spr patch 138366
```
works.

pcc commented:
Sorry, I was having a problem with spr with this change so I pushed the
updated change directly to the branch.

I think there are some merge conflicts by now in the include list but they
should be easy to resolve.
-- 
Peter

On Sat, Jun 14, 2025, 17:58 Fangrui Song ***@***.***> wrote:

> *MaskRay* left a comment (llvm/llvm-project#138366)
> <https://github.com/llvm/llvm-project/pull/138366#issuecomment-2973404563>
>
> pcc <https://github.com/pcc> wants to merge 1 commit into
> users/pcc/spr/main.elf-add-branch-to-branch-optimization
> <https://github.com/llvm/llvm-project/tree/users/pcc/spr/main.elf-add-branch-to-branch-optimization>
> from users/pcc/spr/elf-add-branch-to-branch-optimization
> <https://github.com/llvm/llvm-project/tree/users/pcc/spr/elf-add-branch-to-branch-optimization>
>
> Is the base branch associated with an open PR?
>
> Neither
>
> curl -L https://github.com/llvm/llvm-project/pull/138366.diff | patch -p1
>
> spr patch 138366
>
> works.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/138366#issuecomment-2973404563>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADHYQAZ2GOB3GOPSCGQL3L3DTANHAVCNFSM6AAAAAB4LCEMI2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNZTGQYDINJWGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

MaskRay commented:
Perhaps you'll need to change the base branch to `main` and force push to users/pcc/spr/elf-add-branch-to-branch-optimization?

MaskRay commented on lld/test/ELF/x86-64-branch-to-branch.s:
```diff
@@ -0,0 +1,98 @@
+.globl vtable
+vtable:
+# B2B: Contents of section .rodata:
+# B2B-NEXT: [[VF:[0-9a-f]{8}]]
```
We need to check the exact value, otherwise the PLT32 with an addend not -4 is not tested.

MaskRay commented on lld/ELF/Arch/TargetImpl.h:
```diff
@@ -0,0 +1,94 @@
+//===- TargetImpl.h ---------------------------------------------*- C++ -*-===//
```
`//===----------------------------------------------------------------------===//` for new file per https://llvm.org/docs/CodingStandards.html#file-headers

MaskRay commented on lld/ELF/Arch/TargetImpl.h:
```diff
@@ -0,0 +1,93 @@
+//===- TargetImpl.h -------------------------------------------------------===//
```
Nit: switch to the new style as well

MaskRay commented on lld/ELF/InputSection.cpp:
```diff
@@ -430,8 +430,10 @@ InputSectionBase *InputSection::getRelocatedSection() const {
 void InputSection::copyRelocations(Ctx &ctx, uint8_t *buf) {
-  if (ctx.arg.relax && !ctx.arg.relocatable &&
-      (ctx.arg.emachine == EM_RISCV || ctx.arg.emachine == EM_LOONGARCH)) {
+  if (!ctx.arg.relocatable &&
```
The condition is now complex... Perhaps define a variable for linker relaxation targets (RISCV,LoongArch)?
```
bool linkerRelax = ctx.arg.relax && (is_contained({EM_RISCV, EM_LOONGARCH}, ctx.arg.emachine);
...
(linkerRelax || ctx.arg.branchToBranch)
```

MaskRay commented on lld/docs/ld.lld.1:
```diff
@@ -93,6 +93,10 @@ Bind default visibility defined STB_GLOBAL function symbols locally for
+.It Fl -branch-to-branch
+Enable the branch-to-branch optimizations: a branch whose target is
+another branch instruction is rewritten to point to the latter branch
+target (AArch64 and X86_64 only). Enabled by default at -O2.
```
-O2 should use `.Fl O2.` (unfortunately the dot after `-O2` has to be highlighted as well...)

jrtc27 replied:
`.Fl O2 Ns .`?

MaskRay reviewed (approved):
Looks good to me! Let's give @smithp35 a chance to share the feedback too.

MaskRay commented:
Could you also update lld/docs/ReleaseNotes.rst?

pcc commented:
Merged https://github.com/llvm/llvm-project/commit/491b82a5ec1add78d2c93370580a2f1897b6a364

zmodem commented:
We're hitting an assert after this. I've attached a reproducer at https://crbug.com/427013249#comment3

```
$ tar zxf /tmp/repro.tar.gz
$ ( cd repro && ../build/bin/ld.lld @response.txt )
ld.lld: /work/llvm-project/llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From*) [with To = lld::elf::InputSection; From = lld::elf::InputSectionBase]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ../build/bin/ld.lld @response.txt
 #0 0x000055ef8070ae7b llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (../build/bin/ld.lld+0x132ae7b)
 #1 0x000055ef80707f76 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #2 0x00007f356d359df0 (/lib/x86_64-linux-gnu/libc.so.6+0x3fdf0)
 #3 0x00007f356d3ae95c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #4 0x00007f356d359cc2 raise ./signal/../sysdeps/posix/raise.c:27:6
 #5 0x00007f356d3424ac abort ./stdlib/abort.c:81:3
 #6 0x00007f356d342420 __assert_perror_fail ./assert/assert-perr.c:31:1
 #7 0x000055ef80042c1d (anonymous namespace)::X86_64::applyBranchToBranchOpt() const (.cold) X86_64.cpp:0:0
 #8 0x000055ef80a287e4 (anonymous namespace)::Writer<llvm::object::ELFType<(llvm::endianness)1, true>>::finalizeSections() Writer.cpp:0:0
 #9 0x000055ef80a2a3f0 void lld::elf::writeResult<llvm::object::ELFType<(llvm::endianness)1, true>>(lld::elf::Ctx&) (../build/bin/ld.lld+0x164a3f0)
#10 0x000055ef80884a35 void lld::elf::LinkerDriver::link<llvm::object::ELFType<(llvm::endianness)1, true>>(llvm::opt::InputArgList&) (../build/bin/ld.lld+0x14a4a35)
#11 0x000055ef8088bfe2 lld::elf::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) (../build/bin/ld.lld+0x14abfe2)
#12 0x000055ef8088c8ff lld::elf::link(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, bool, bool) (../build/bin/ld.lld+0x14ac8ff)
#13 0x000055ef8074349a lld::unsafeLldMain(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, llvm::ArrayRef<lld::DriverDef>, bool) (../build/bin/ld.lld+0x136349a)
#14 0x000055ef80681409 lld_main(int, char**, llvm::ToolContext const&) (../build/bin/ld.lld+0x12a1409)
#15 0x000055ef80532643 main (../build/bin/ld.lld+0x1152643)
#16 0x00007f356d343ca8 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#17 0x00007f356d343d65 call_init ./csu/../csu/libc-start.c:128:20
#18 0x00007f356d343d65 __libc_start_main ./csu/../csu/libc-start.c:347:5
#19 0x000055ef80680ce1 _start (../build/bin/ld.lld+0x12a0ce1)
Aborted (core dumped)
```

I'll revert this for now.

pcc commented:
> We're hitting an assert after this. I've attached a reproducer at https://crbug.com/427013249#comment3
> 
> ```
> $ tar zxf /tmp/repro.tar.gz
> $ ( cd repro && ../build/bin/ld.lld @response.txt )
> ld.lld: /work/llvm-project/llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From*) [with To = lld::elf::InputSection; From = lld::elf::InputSectionBase]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
> PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
> Stack dump:
> 0.      Program arguments: ../build/bin/ld.lld @response.txt
>  #0 0x000055ef8070ae7b llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (../build/bin/ld.lld+0x132ae7b)
>  #1 0x000055ef80707f76 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
>  #2 0x00007f356d359df0 (/lib/x86_64-linux-gnu/libc.so.6+0x3fdf0)
>  #3 0x00007f356d3ae95c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
>  #4 0x00007f356d359cc2 raise ./signal/../sysdeps/posix/raise.c:27:6
>  #5 0x00007f356d3424ac abort ./stdlib/abort.c:81:3
>  #6 0x00007f356d342420 __assert_perror_fail ./assert/assert-perr.c:31:1
>  #7 0x000055ef80042c1d (anonymous namespace)::X86_64::applyBranchToBranchOpt() const (.cold) X86_64.cpp:0:0
>  #8 0x000055ef80a287e4 (anonymous namespace)::Writer<llvm::object::ELFType<(llvm::endianness)1, true>>::finalizeSections() Writer.cpp:0:0
>  #9 0x000055ef80a2a3f0 void lld::elf::writeResult<llvm::object::ELFType<(llvm::endianness)1, true>>(lld::elf::Ctx&) (../build/bin/ld.lld+0x164a3f0)
> #10 0x000055ef80884a35 void lld::elf::LinkerDriver::link<llvm::object::ELFType<(llvm::endianness)1, true>>(llvm::opt::InputArgList&) (../build/bin/ld.lld+0x14a4a35)
> #11 0x000055ef8088bfe2 lld::elf::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) (../build/bin/ld.lld+0x14abfe2)
> #12 0x000055ef8088c8ff lld::elf::link(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, bool, bool) (../build/bin/ld.lld+0x14ac8ff)
> #13 0x000055ef8074349a lld::unsafeLldMain(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, llvm::ArrayRef<lld::DriverDef>, bool) (../build/bin/ld.lld+0x136349a)
> #14 0x000055ef80681409 lld_main(int, char**, llvm::ToolContext const&) (../build/bin/ld.lld+0x12a1409)
> #15 0x000055ef80532643 main (../build/bin/ld.lld+0x1152643)
> #16 0x00007f356d343ca8 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
> #17 0x00007f356d343d65 call_init ./csu/../csu/libc-start.c:128:20
> #18 0x00007f356d343d65 __libc_start_main ./csu/../csu/libc-start.c:347:5
> #19 0x000055ef80680ce1 _start (../build/bin/ld.lld+0x12a0ce1)
> Aborted (core dumped)
> ```
> 
> I'll revert this for now.

Thanks, sent #145579 with the fix.
